analysis/priestdiscipline/src/modules/spells/Penance.js

```diff
@@ -48,19 +31,35 @@ class Penance extends Analyzer {
       return;
     }
 
-    this.eventGrouper.processEvent(event);
-
-    event.penanceBoltNumber = this.currentBoltNumber;
+    event.penanceBoltNumber = this._boltCount;
+    this._boltCount += 1;
   }
 
   onHeal(event) {
     if (!Penance.isPenance(event.ability.guid)) {
       return;
     }
 
-    this.eventGrouper.processEvent(event);
+    event.penanceBoltNumber = this._boltCount;
+    this._boltCount += 1;
+  }
+
+  onCast(event) {
+    this._casts += 1;
+    this._missedBolts += this._expectedBoltCount - this._boltCount;
+    this.calculateExpectedBolts(event);
+    this._boltCount = 0;
+  }
 
-    event.penanceBoltNumber = this.currentBoltNumber;
+  calculateExpectedBolts(cast = null) {
```

Comment:
Only one function calls this and both are pretty simple in terms of total lines of code so we should probably inline this. 
IE moving the `if(this.selectedCombatant.hasBuff...` into this function